x/examples/soax-relay/main.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/netip"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	csocks5 "github.com/Jigsaw-Code/outline-sdk/transport/socks5"
+	"github.com/spf13/viper"
+	"github.com/things-go/go-socks5"
+	"github.com/things-go/go-socks5/statute"
+)
+
+// CustomAuthenticator handles authentication based on username prefix and extracts the suffix
+type CustomAuthenticator struct {
+	PrefixCredentials CredentialStore // Store for validating the prefix part of the username
+}
+
+// GetCode implements the `GetCode` method for the `Authenticator` interface
+func (a CustomAuthenticator) GetCode() uint8 {
+	return statute.MethodUserPassAuth
+}
+
+// Authenticate checks if the username matches a required prefix and extracts the suffix
+func (a CustomAuthenticator) Authenticate(reader io.Reader, writer io.Writer, userAddr string) (*socks5.AuthContext, error) {
+	// Respond to the client to use username/password authentication
+	if _, err := writer.Write([]byte{statute.VersionSocks5, statute.MethodUserPassAuth}); err != nil {
+		return nil, err
+	}
+
+	// Parse the username and password from the client request
+	nup, err := statute.ParseUserPassRequest(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	fullUsername := string(nup.User)
+	password := string(nup.Pass)
+
+	// Split the username into prefix and suffix parts
+	parts := strings.SplitN(fullUsername, "-", 2)
+	if len(parts) < 2 {
+		if _, err := writer.Write([]byte{statute.UserPassAuthVersion, statute.AuthFailure}); err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("username does not contain a valid prefix-suffix format")
+	}
+
+	usernamePrefix := parts[0] // Prefix (e.g., "jane")
+	usernameSuffix := parts[1] // Suffix (e.g., "-country-ir")
+
+	log.Printf("Attempting authentication with prefix: %s and suffix: %s", usernamePrefix, usernameSuffix)
+
+	// Validate the prefix and password against stored credentials
+	if !a.PrefixCredentials.Valid(usernamePrefix, password, userAddr) {
```

Comment:
I would simply do the map lookup here to keep it simple. This is not reusable code anyway.